### PR TITLE
Document & test multiple themes

### PR DESCRIPTION
**Pull request by mhutter**
Follow-up to #79

This commit adds support for mounting the same volume multiple times, required for the "multiple themes" use case.

Since simply mounting the themes volume to `/opt/jboss/keycloak/themes` would overwrite the default/builtin themes, multiple mounts are required.

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

### CODE REVIEW

File: class/defaults.yml [hunk 1]
```diff
@@ -88,7 +88,7 @@ parameters:
     # Extra volumes that are appended to the helm variable `extraVolumes`
     extraVolumes: {}
     # Extra volume mounts that are appended to the helm variable `extraVolumeMounts`
-    extraVolumeMounts: {}
+    extraVolumeMounts: []
```
Comment (ccremer) [inline class/defaults.yml:91]: This looks like a breaking change. Please add a how-to page that instructs how to upgrade to the next major version.

File: docs/modules/ROOT/pages/how-tos/custom-theme.adoc [hunk 2]
```diff
@@ -40,25 +40,33 @@ parameters:
...
-        readOnly: true
+      - name: theme
         mountPath: /opt/jboss/keycloak/themes/company
+        subPath: company
+        readOnly: true
+      - name: theme
+        readOnly: true
+        mountPath: /opt/jboss/keycloak/themes/other-theme
+        subPath: other-theme
+        readOnly: true
```
Comment (ccremer) [inline docs/modules/ROOT/pages/how-tos/custom-theme.adoc:69]: Is this page up-to-date? Somehow the volumeMount `themes` doesn't match `theme` in the `extraVolumeMounts` array?

**Review by ccremer — CHANGES_REQUESTED** [review]

**Review by glrf — APPROVED** [review]
LGTM apart from the inconsistency of `theme` and `themes`.

Also to be cleaner and consistent an upgrade guide is a good idea, even if you update all instances yourself.

**Comment by mhutter** [comment]
Agree, will add a how-to page.

I also considered implementing the change to be backwards compatible (by converting structs to arrays), but this would only hide deprecated configs and make it more painful later :)

File: class/defaults.yml (hunk 1 above)
Comment (ccremer) [reply class/defaults.yml:91]: @mhutter actually, how does this work with merging from hierarchy? Converting to array would remove the possibility to define volume mounts from commodore defaults yet still add mounts specific to clusters? Or is still possible?

**Comment by glrf** [comment]
I might just have had a better idea to solve this without a breaking change and without losing the option to overwrite defaults.

We could keep the map and add an optional field to specify the volume. So:


```
extraVolumeMounts:
  test:
    readOnly: true
    mountPath: /opt/test
  testToo:
    name: test
    readOnly: true
    mountPath: /opt/testToo

```

Would result in 

```
- name: test
  readOnly: true
  mountPath: /opt/test
- name: test
  readOnly: true
  mountPath: /opt/testToo
```

**Comment by ccremer** [comment]
> I might just have had a better idea to solve this without a breaking change and without losing the option to overwrite defaults.
> 
> We could keep the map and add an optional field to specify the volume. So:
> 
> ```
> extraVolumeMounts:
>   test:
>     readOnly: true
>     mountPath: /opt/test
>   testToo:
>     name: test
>     readOnly: true
>     mountPath: /opt/testToo
> ```
> 
> Would result in
> 
> ```
> - name: test
>   readOnly: true
>   mountPath: /opt/test
> - name: test
>   readOnly: true
>   mountPath: /opt/testToo
> ```

Yes, I thought of something similar. The whole idea of #67 was to enable volume mounting from different places in the hierarchy through deep-merge. By converting to array we would loose this possibility and this feels a step backwards.
Remember, volume mounts are not just needed for themes, but for custom Keycloak extensions as well (loading JAR files).

File: class/defaults.yml (hunk 1 above)
Comment (mhutter) [reply class/defaults.yml:91]: Not sure, but what is the use case for hierarchical configuration of volume mounts?

**Comment by mhutter** [comment]
Yes, i thought of this as well, converting the key in the dict to a kind of "id" instead.

It would also allow to modify existing volumeMounts further down in the hierarchy.

File: class/defaults.yml (hunk 1 above)
Comment (ccremer) [reply class/defaults.yml:91]: One use case would be:
- Define a common theme in the tenant
- Define an additional init container that uses volume mounts for a specific cluster in order to test a Dev build of keycloak extensions

Both themeing and extension make use of volume mounts

File: docs/modules/ROOT/pages/how-tos/custom-theme.adoc (hunk 2 above)
Comment (mhutter) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc]: Updated the docs to be more consistent. I used multiple themes by default since it's easier to deduct how a single theme works from there

File: docs/modules/ROOT/pages/how-tos/custom-theme.adoc [hunk 3]
```diff
@@ -40,27 +40,39 @@ parameters:
...
     extraVolumes:
-      theme:
+      themes:
         emptyDir: {}
+
     extraVolumeMounts:
-      theme:
-        readOnly: true
+      theme-company:
+        name: themes
```
Comment (ccremer) [inline docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: I don't see any jsonnet changes in the postprocessing that makes use of the new parameter `name`.
Do you intend to do a separate PR for this?

Comment (mhutter) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: No, as it turns out everything was already fixed in my previous PR, making this one a documentation PR only :)

Comment (mhutter) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: Oh, and of course the tests that verify the intended behavior

Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: I was kinda skipping the tests :)
Please help me understand: Is the name parameter actually required now? That would make #79 a breaking change. Should we go for a non-breaking change if we default to the key name if the `name` parameter is missing?

Comment (glrf) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: As far as I understood, the following will overwrite the `name` if it is present and otherwise use the key `k`. 

https://github.com/projectsyn/component-keycloak/blob/450ea1f79f9279d5d65b3c9144b40821325e1376/postprocess/extra-env.jsonnet#L13-L18

**Review by glrf — APPROVED** [review]

File: docs/modules/ROOT/pages/how-tos/custom-theme.adoc (hunk 3 above)
Comment (mhutter) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: Exactly, `name` wil be set to the key but overwritten if explicitly set. This was actually the case all along, but before #79 the mounts were deduplicated based on their `name` attribute (which is wrong since `name` identifies the volume, and not the volumeMount).

Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/custom-theme.adoc:62]: thanks for clearing that up.

**Review by ccremer — APPROVED** [review]
